dbcon/joblist/tupleannexconstantstep.cpp

```diff
@@ -0,0 +1,563 @@
+          {
+            rowGroupSkip.incRowCount();
+            rowSkip.nextRow();
+            if (UNLIKELY(rowGroupSkip.getRowCount() >= rowgroup::rgCommonSize))
+            {
+              // allocate new RGData for skipped rows below the fLimitStart
+              // offset (do not take it into account in RM assuming there
+              // are few skipped rows
+              dataVecSkip.push_back(rgDataSkip);
+              rgDataSkip.reinit(rowGroupSkip);
+              rowGroupSkip.setData(&rgDataSkip);
+              rowGroupSkip.resetRowGroup(0);
+              rowGroupSkip.getRow(0, &rowSkip);
+            }
+            continue;
+          }
+
+          ++fRowsReturned;
+
+          fRowGroupOut.incRowCount();
+          fRowOut.nextRow();
+
+          if (UNLIKELY(fRowGroupOut.getRowCount() >= rowgroup::rgCommonSize))
+          {
+            dataVec.push_back(rgDataOut);
+            rgDataOut.reinit(fRowGroupOut);
+            fRowGroupOut.setData(&rgDataOut);
+            fRowGroupOut.resetRowGroup(0);
+            fRowGroupOut.getRow(0, &fRowOut);
+          }
+        }
+      }
+
+      more = fInputDL->next(fInputIterator, &rgDataIn);
+    }
+
+    if (fRowGroupOut.getRowCount() > 0)
+      dataVec.push_back(rgDataOut);
+
+    for (vector<RGData>::iterator i = dataVec.begin(); i != dataVec.end(); i++)
+    {
+      rgDataOut = *i;
+      fRowGroupOut.setData(&rgDataOut);
+      fOutputDL->insert(rgDataOut);
+    }
+  }
+  catch (...)
+  {
+    handleException(std::current_exception(), logging::ERR_IN_PROCESS, logging::ERR_ALWAYS_CRITICAL,
+                    "TupleAnnexConstantStep::executeWithDistinct()");
+  }
+
+  while (more)
+    more = fInputDL->next(fInputIterator, &rgDataIn);
+
+  // Bug 3136, let mini stats to be formatted if traceOn.
+  fOutputDL->endOfInput();
+}
+
+
+void TupleAnnexConstantStep::executeNoDistinct()
+{
+  utils::setThreadName("TACS");
+  RGData rgDataIn;
+  RGData rgDataOut;
+  bool more = false;
+
+  try
+  {
+    more = fInputDL->next(fInputIterator, &rgDataIn);
+
+    if (traceOn())
+      dlTimes.setFirstReadTime();
+
+    StepTeleStats sts;
+    sts.query_uuid = fQueryUuid;
+    sts.step_uuid = fStepUuid;
+    sts.msg_type = StepTeleStats::ST_START;
+    sts.total_units_of_work = 1;
+    postStepStartTele(sts);
+
+    while (more && !cancelled() && !fLimitHit)
+    {
+      fRowGroupIn.setData(&rgDataIn);
+      fRowGroupIn.getRow(0, &fRowIn);
+      // Get a new output rowgroup for each input rowgroup to preserve the rids
+      rgDataOut.reinit(fRowGroupOut, fRowGroupIn.getRowCount());
+      fRowGroupOut.setData(&rgDataOut);
+      fRowGroupOut.resetRowGroup(fRowGroupIn.getBaseRid());
+      fRowGroupOut.setDBRoot(fRowGroupIn.getDBRoot());
+      fRowGroupOut.getRow(0, &fRowOut);
+
+      for (uint64_t i = 0; i < fRowGroupIn.getRowCount() && !cancelled() && !fLimitHit; ++i)
+      {
+        // skip first limit-start rows
+        if (fRowsProcessed++ < fLimitStart)
```

Comment:
This is for limit processing and this step doesn't have it.

Reply:
Removed.